### PR TITLE
feat: KServe Bindings to start tritonfrontend

### PR DESCRIPTION
**Pull request by KrishnanPrash**
Adding support to start the `KServeHttp` and `KServeGrpc` frontends from the `server/python/openai/openai_frontend/main.py`

### CODE REVIEW

File: python/openai/openai_frontend/main.py [hunk 1]
```diff
@@ -28,12 +28,24 @@
...
+import sys
 from functools import partial
 
 import tritonserver
 from engine.triton_engine import TritonLLMEngine
 from frontend.fastapi_frontend import FastApiFrontend
 
+try:
+    from tritonfrontend import KServeGrpc, KServeHttp
+except:
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply python/openai/openai_frontend/main.py:40]: except ImportError ?

File: python/openai/openai_frontend/main.py [hunk 2]
```diff
@@ -104,6 +129,17 @@ def main():
         log_error=True,
     ).start(wait_until_ready=True)
 
+    http_options, http_service = None, None
+    grpc_options, grpc_service = None, None
+    if "tritonfrontend" in sys.modules:
```
Comment (GuanLuo) [inline python/openai/openai_frontend/main.py:134]: defer import to here?

File: python/openai/openai_frontend/main.py [hunk 3]
```diff
@@ -28,12 +28,24 @@
...
 
+try:
+    from tritonfrontend import KServeGrpc, KServeHttp
+except:
+    print(
+        """
+          tritonfrontend is not present in this environment.
+          Check /opt/tritonserver/python for tritonfrontend*.whl/
+          If present, please pip install.
+          """
```
Comment (rmccorm4) [inline python/openai/openai_frontend/main.py]: This prints with some weird spacing. Can we print something like this?
```python
print(
    "[WARNING] The 'tritonfrontend' package was not found. "
    "KServe frontends won't be available through this application without it. "
    "Check /opt/tritonserver/python for tritonfrontend*.whl and pip install it if present."
)
```

File: python/openai/openai_frontend/main.py [hunk 4]
```diff
@@ -104,6 +118,28 @@ def main():
...
+        grpc_service.start()
+
+    except ModuleNotFoundError:
+        print(
+            """
+          tritonfrontend is not present in this environment.
+          Check /opt/tritonserver/python for tritonfrontend*.whl/
+          If present, please pip install.
+          """
+        )
```
Comment (rmccorm4) [inline python/openai/openai_frontend/main.py:141]: Use a helper function, ex: `http_service, grpc_service = start_kserve_frontends(server, args)`

File: python/openai/openai_frontend/main.py (hunk 2 above)
Comment (kthui) [reply python/openai/openai_frontend/main.py]: Does it work as intended by moving the import there?
```
Python 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> "numpy" in sys.modules
False
>>> import numpy
>>> "numpy" in sys.modules
True
>>>
```

File: python/openai/openai_frontend/main.py [hunk 5]
```diff
@@ -52,7 +53,7 @@ def parse_args():
     # Uvicorn
     uvicorn_group = parser.add_argument_group("Uvicorn")
     uvicorn_group.add_argument("--host", type=str, default=None, help="host name")
-    uvicorn_group.add_argument("--port", type=int, default=8000, help="port number")
+    uvicorn_group.add_argument("--port", type=int, default=9000, help="port number")
```
Comment (rmccorm4) [inline python/openai/openai_frontend/main.py:56]:
```suggestion
    uvicorn_group.add_argument("--port", type=int, default=9000, help="OpenAI port")
```

File: python/openai/openai_frontend/main.py [hunk 6]
```diff
@@ -88,6 +89,19 @@ def parse_args():
         choices=["vllm", "tensorrtllm"],
         help="Manual override of Triton backend request format (inputs/output names) to use for inference",
     )
+    # Should this be wrapped in an if block that only enters if import tritonfrontend worked
+    triton_group.add_argument(
+        "--kserve-http-port",
+        type=int,
+        default=8000,
+        help="Triton KServeHTTP port number",
```
Comment (rmccorm4) [inline python/openai/openai_frontend/main.py:97]:
```suggestion
        help="KServe Predict v2 HTTP port",
```

File: python/openai/openai_frontend/main.py [hunk 7]
```diff
@@ -88,6 +89,19 @@ def parse_args():
...
+    )
+    triton_group.add_argument(
+        "--kserve-grpc-port",
+        type=int,
+        default=8001,
+        help="Triton KServeGrpc port number",
```
Comment (rmccorm4) [inline python/openai/openai_frontend/main.py:103]:
```suggestion
        help="KServe Predict v2 GRPC port (default: 8001)",
```

File: python/openai/openai_frontend/main.py (hunk 2 above)
Comment (KrishnanPrash) [reply python/openai/openai_frontend/main.py]: I was planning on replacing the logic to look something like this:
```python3
http_service, grpc_service = None, None
try:
     from tritonfrontend import KServeGrpc, KServeHttp
     http_service, grpc_service = start_kserve_frontends(server, args)
except ModuleNotFoundError:
     print(...)
     
# OpenAI Frontend logic
# Blocking call until killed or interrupted with SIGINT
frontend.start()

if http_service:
    http_service.stop()
if grpc_service:
    grpc_service.stop()
```

File: python/openai/openai_frontend/main.py (hunk 3 above)
Comment (KrishnanPrash) [reply python/openai/openai_frontend/main.py]: New output formatting:
image
Thank you for suggestion!

**Review by kthui — APPROVED** [review]
Nice work! Please make sure @rmccorm4 is good with the change before merging.

File: python/openai/openai_frontend/main.py [hunk 8]
```diff
@@ -45,14 +45,36 @@ def signal_handler(server, frontend, signal, frame):
...
+
 def parse_args():
     parser = argparse.ArgumentParser(
         description="Triton OpenAI Compatible RESTful API server."
     )
     # Uvicorn
     uvicorn_group = parser.add_argument_group("Uvicorn")
     uvicorn_group.add_argument("--host", type=str, default=None, help="host name")
-    uvicorn_group.add_argument("--port", type=int, default=8000, help="port number")
+    uvicorn_group.add_argument("--port", type=int, default=9000, help="OpenAI port")
```
Comment (rmccorm4) [inline python/openai/openai_frontend/main.py:77]:
```suggestion
    uvicorn_group.add_argument("--openai-port", type=int, default=9000, help="OpenAI HTTP port (default: 9000)")
```

**Comment by rmccorm4** [comment]
LGTM - but can you update the `README.md` to replace port 8000 with port 9000 for OpenAI call examples?

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**